src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs

```diff
@@ -3402,6 +3402,44 @@ public async Task UploadStreamReleasesHubActivatorOnceComplete()
             }
         }
 
+        [Fact]
+        public async Task UploadStreamFromSendReleasesHubActivatorOnceComplete()
+        {
+            using (StartVerifiableLog())
+            {
+                var serviceProvider = HubConnectionHandlerTestUtils.CreateServiceProvider(builder =>
+                {
+                    builder.AddSingleton(typeof(IHubActivator<>), typeof(CustomHubActivator<>));
+                }, LoggerFactory);
+                var connectionHandler = serviceProvider.GetService<HubConnectionHandler<MethodHub>>();
+
+                using (var client = new TestClient())
+                {
+                    var connectionHandlerTask = await client.ConnectAsync(connectionHandler).OrTimeout();
+
+                    await client.BeginUploadStreamAsync(invocationId: null, nameof(MethodHub.UploadDoesWorkOnComplete), streamIds: new[] { "id" }, args: Array.Empty<object>()).OrTimeout();
+
+                    await client.SendHubMessageAsync(new StreamItemMessage("id", "hello")).OrTimeout();
+                    await client.SendHubMessageAsync(new StreamItemMessage("id", " world")).OrTimeout();
+                    await client.SendHubMessageAsync(CompletionMessage.Empty("id")).OrTimeout();
+
+                    // This task completes if the upload stream is completed, via closing the connection
+                    var task = (Task<int>)client.Connection.Items[nameof(MethodHub.UploadDoesWorkOnComplete)];
+                    await task.OrTimeout();
+
+                    var hubActivator = serviceProvider.GetService<IHubActivator<MethodHub>>() as CustomHubActivator<MethodHub>;
+
+                    // OnConnectedAsync and UploadDoesWorkOnComplete hubs have been disposed
+                    Assert.Equal(2, hubActivator.ReleaseCount);
```

Comment:
This test seems to be flaky. I don't think waiting for UploadDoesWorkOnComplete to exit is waiting long enough, since the the associated scope might not yet have completed.

> Assert.Equal() Failure\nExpected: 2\nActual: 1

https://dev.azure.com/dnceng/public/_build/results?buildId=342272&view=ms.vss-test-web.build-test-results-tab&runId=10146750&resultId=101346&paneView=debug